js/src/util/tokens/index.js

```diff
@@ -0,0 +1,293 @@
+  const imagesPromise = fetchTokensImages(api, tokenReg, tokenIndexes);
+
+  let results;
+
+  return Promise.all(calls)
+    .then((_results) => {
+      results = _results;
+      return imagesPromise;
+    })
+    .then((images) => {
+      return results.map((rawTokenData, index) => {
+        const tokenIndex = tokenIndexes[index];
+        const tokenData = tokenReg.instance.token
+          .decodeOutput(rawTokenData)
+          .map((t) => t.value);
+
+        const [ address, tag, format, name ] = tokenData;
+        const image = images[index];
+
+        const token = {
+          address,
+          id: sha3(address + tokenIndex).slice(0, 10),
+          index: tokenIndex,
+
+          format: format.toString(),
+          image,
+          name,
+          tag,
+
+          fetched: true
+        };
+
+        return token;
+      });
+    });
+}
+
+export function fetchTokensImages (api, tokenReg, tokenIndexes) {
+  const requests = tokenIndexes.map((tokenIndex) => {
+    const metaCalldata = tokenReg.getCallData(tokenReg.instance.meta, {}, [tokenIndex, 'IMG']);
+
+    return { to: tokenReg.address, data: metaCalldata };
+  });
+
+  const calls = requests.map((req) => api.eth.call(req));
+
+  return Promise.all(calls)
+    .then((results) => {
+      return results.map((rawImage) => {
+        const image = tokenReg.instance.meta.decodeOutput(rawImage)[0].value;
+
+        return hashToImageUrl(image);
+      });
+    });
+}
+
+/**
+ * `updates` should be in the shape:
+ *   {
+ *     [ who ]: [ tokenId ]  // Array of tokens to updates
+ *   }
+ *
+ * Returns a Promise resolved with the balances in the shape:
+ *   {
+ *     [ who ]: { [ tokenId ]: BigNumber } // The balances of `who`
+ *   }
+ */
+export function fetchAccountsBalances (api, tokens, updates) {
+  const accountAddresses = Object.keys(updates);
+
+  // Updates for the ETH balances
+  const ethUpdates = accountAddresses
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId === ETH_TOKEN.id);
+    })
+    .reduce((nextUpdates, tokenIds, accountIndex) => {
+      if (tokenIds.length > 0) {
+        const accountAddress = accountAddresses[accountIndex];
+
+        nextUpdates[accountAddress] = tokenIds;
+      }
+
+      return nextUpdates;
+    }, {});
+
+  // Updates for Tokens balances
+  const tokenUpdates = Object.keys(updates)
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId !== ETH_TOKEN.id);
+    })
+    .reduce((nextUpdates, tokenIds, accountIndex) => {
+      if (tokenIds.length > 0) {
+        const accountAddress = accountAddresses[accountIndex];
+
+        nextUpdates[accountAddress] = tokenIds;
+      }
+
+      return nextUpdates;
+    }, {});
+
+  let ethBalances = {};
+  let tokensBalances = {};
+
+  const ethPromise = fetchEthBalances(api, Object.keys(ethUpdates))
+    .then((_ethBalances) => {
+      ethBalances = _ethBalances;
```

Comment:
```js
const ethPromise = fetchEthBalances(api, ...);
const tokenPromise = Object.keys(tokenUpdates).reduce((promise, address) => {
  return promise.then(balances => {
     return fetchTokensBalances(api, updateTokens, ...).then(newBalances => {
       balances[address] = newBalances[address];
       return balances;
     });
  });
}, Promise.resolve({}));

return Promise.all([ethPromise, tokenPromise]).then(([ethBalances, tokensBalances]) => {
 ...
});
```

Reply:
Now that I wrote the code I think there is a bit of redundancy in calling `fetchTokensBalances`:

Why can't we do a single query for all tokens seen in any account and all accounts at once?
Something like:
```js
const updateTokens = uniq(Object.values(tokenUpdates).reduce((all, ids) => all.concat(ids), []));
const accounts = Object.keys(tokenUpdates);
const tokenPromise = fetchTokensBalances(api, updateTokens, accounts);
```
?

Reply:
So it's because not all accounts need to update the same tokens.

Reply:
Yeah I agree that it's hard to read, and the whole thing could be simplified if we didn't really care about having a fine-tuned fetching. But I guess the scope of the PR was to be able to make _just_ the right number of calls.